core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -525,6 +546,31 @@ public void refresh(String source) throws EngineException {
     }
 
     @Override
+    public void addRefreshListener(RefreshListener listener) {
+        requireNonNull(listener, "listener cannot be null");
+        Translog.Location listenerLocation = requireNonNull(listener.location(), "listener's location cannot be null");
+
+        Translog.Location lastRefresh = lastRefreshedLocation;
+        if (lastRefresh != null && lastRefresh.compareTo(listenerLocation) >= 0) {
+            // Location already visible, just call the listener
+            listener.refreshed(false);
+            return;
+        }
+        if (refreshListenersEstimatedSize < engineConfig.getIndexSettings().getMaxRefreshListeners()) {
```

Comment:
can we just use a non-concurrent datastructure and make this synchronized. Maybe we can just use a simple ArrayList  with preallocated space and be done with it? I don't think we need this estimated size here we can just make this simple and move into a synced block?